customerio/customerio-ios
# chore: animate height changes inline in-app SwiftUI View

Ticket: https://linear.app/customerio/issue/MBL-402/

# Problem 

The SwiftUI inline in-app message View instantly jumps from height 0 to N height when an in-app message is rendered. We need a smooth animation when the height changes like we have in the UIKit inline View. 

Besides this, the SwiftUI view currently does not do anything when the close button is pressed. 

This PR solves all of these problems by animating between all of the states. 

# Solution 

Originally, I thought that the SwiftUI inline View implementation was going to be a very small wrapper around the UIKit inline View. I was thinking that if the UIKit inline View implementation would animate height changes, SwiftUI would automatically be able to display those UIKit animations, too. 

I heave learned that when using `UIViewRepresentable`, SwiftUI adds auto layout constraints that you do not control. These SwiftUI auto layout constraints are not compatible with the UIKit inline View height animations that are implemented. 

To get around this issue, we can make an architecture change. We have a SwiftUI implementation and a separate UIKit implementation where each of them implement their own animation logic: 

Architecture before this PR: 
```
SwiftUI View (UIViewRepresentable)
  \__ UIKit View <--- performs animations when height changes
    \__ UIKit WebView renders in-app message
```

Architecture after this PR: 
```
SwiftUI View <--- performs animations when height changes
  \__ UIKit WebView renders in-app message

UIKit View <--- performs animations when height changes 
  \__ UIKit WebView renders in-app message
```

This PR is the SwiftUI View that performs animations when height of WebView changes. A follow-up PR will be made for the UIKit View.

# Testing 

* Open Dashboard in SwiftUI sample app. Send an in-app message. Expect inline View to be height 0 when first displayed. 
* When in-app message renders, animate height to N to display the in-app message. 
* Click close button. If there is another message in local queue, animate the height change to the height of next message. 
* Click close button. If no other message in local queue, animate height to 0 to dismiss in-app message. 

Demo video:

https://github.com/user-attachments/assets/3cdfa23e-06da-4baa-8512-f654b87e292e
